### PR TITLE
[AOT] Name mangling in AOT

### PR DESCRIPTION
**Pull request by giuseros**
Mini-RFC is here: https://discuss.tvm.apache.org/t/mini-rfc-name-mangling-in-aot

With this change we'll mangle the name of global symbols so that we can bundle
together multiple models in the same application.

The relay.build interface has been left unchanged, which means I am
resuing mod_name as a prefix for all functions. 
* If mod_name is None then `tvmgen_` prefix is used
* If mod_name is specified then `tvmgen_mod_name_` prefix is used

I refactored the aot test utils and added some tests for multiple
models.

Change-Id: I310af75c24e422861aeaceb3c3cd4cd602071df5

### CODE REVIEW

File: python/tvm/micro/model_library_format.py
Comment (giuseros) [reply python/tvm/micro/model_library_format.py:58]: The idea here is that if we can specify the name also for the files, I can have those files into the same folder. This makes testing (and hence deployment) way, way easier. About the naming, let's discuss it on the RFC. I don't have any strong opinion, but the idea here was that if we do `{model_name}_function` for the operators we can also do `{model_name}_lib.c` for the file names. If we decide to put `_tvm_ as a constant prefix, then let's put it also in front of the file names (without the `_`)

File: python/tvm/relay/build_module.py
Comment (giuseros) [reply python/tvm/relay/build_module.py:153]: I can't use the default, unfortunately. If I specify `None` the object system will have issues to convert `None` to an empty string. That's why I need to construct the empty string manually.

**Comment by giuseros** [comment]
I made the name-mangling logic completely decoupled by the non-name-mangling logic. This means that name mangling is enabled only in AOT and in `PartitionGraphWithModName`, the rest of TVM won't be using this feature by default.

File: src/relay/transforms/partition_graph.cc
Comment (areusch) [inline src/relay/transforms/partition_graph.cc:526]: it seems like it may be a better design to create a "mangle" pass. curious what your thoughts are on this?

**Review by areusch — CHANGES_REQUESTED** [review]
hi @giuseros, thanks for adding this! here's an initial round of review comments

File: src/target/source/codegen_c_host.cc
Comment (giuseros) [reply src/target/source/codegen_c_host.cc:396]: Do you mean adding an attribute to the TIR function, e.g., `runner_func` and then this would become 
`const bool is_aot_executor_fn = (kv.first->GetAttr("runner_func") == true)`?

File: python/tvm/relay/build_module.py
Comment (giuseros) [reply python/tvm/relay/build_module.py:121]: I didn't think about any restrictions. The idea was to use the name provided by the user as it is. You are asking something like : `verify("&%!") == false`? If yes, two points:
1. I would only validate. If the name is not valid, an error should be thrown
2. Any suggestions on what makes a "valid" name?

File: tests/python/relay/aot/aot_test_utils.py
Comment (giuseros) [reply tests/python/relay/aot/aot_test_utils.py:38]: It's not used anywhere except for those tests. I would move this to be a util function only when (and if) we need it.

File: python/tvm/micro/model_library_format.py
Comment (giuseros) [reply python/tvm/micro/model_library_format.py:65]: Mmmm... Not sure I follow :) To be clear, I am adding a prefix to the lib names so that multiple source files can be in the same folder. Otherwise the user would need to compile each library and manually copy/paste the `lib*.c` files in a common folder with different names.

File: src/target/source/codegen_c_host.cc
Comment (areusch) [reply src/target/source/codegen_c_host.cc:396]: yes, although would suggest to include "aot_" in the attr name and create a constant

File: python/tvm/relay/build_module.py
Comment (areusch) [reply python/tvm/relay/build_module.py:121]: cool, i agree with that. mangling is annoying because you have to be 1:1 and doing this may confuse the user. filesystems can handle a lot of strange filenames now--I think C function names are the more restrictive use case of `mod_name` now. Looks like the spec is digits, underscores, letters, and escaped unicode chars. I'd suggest allowing all of that except the escaped unicode. we don't have the restriction on the first char being a digit since we will prefix this.

File: python/tvm/micro/model_library_format.py
Comment (areusch) [reply python/tvm/micro/model_library_format.py:65]: yeah i like this idea, but we should bump the [version](https://github.com/apache/tvm/blob/main/python/tvm/micro/model_library_format.py#L225) in the metadata to indicate a format change. since MLF is new and not used much yet, i don't think this is consequential, and would rather get in the practice of doing this now rather than trying to judge some more confusing criteria later on.

File: src/relay/transforms/partition_graph.cc
Comment (giuseros) [reply src/relay/transforms/partition_graph.cc]: Adding the `PartitionGraphWithModName` is a very simple change, but maybe is a tiny less elegant than a `MangleGraph` pass (which is a bit more code to write, do to a simple thing).I implemented (and used) the `MangleGraph` pass, so that you can see the difference. I would be still go for a `PartitionGraphWithModName`  but I don't have a strong opinion here.

File: include/tvm/runtime/module.h
Comment (manupa-arm) [inline include/tvm/runtime/module.h:234]: I think we should drop "func" because it adds little value as a name of the function.

cc : @areusch

File: python/tvm/micro/model_library_format.py
Comment (manupa-arm) [inline python/tvm/micro/model_library_format.py:61]: Maybe we can make the default a empty string and append it always ?

File: python/tvm/relay/build_module.py
Comment (manupa-arm) [inline python/tvm/relay/build_module.py:84]: Why do we need this to be true if mod_name is None ?

File: python/tvm/relay/build_module.py
Comment (manupa-arm) [reply python/tvm/relay/build_module.py:167]: Why cant the default be the empty string itself?

File: python/tvm/relay/transform/transform.py
Comment (manupa-arm) [inline python/tvm/relay/transform/transform.py:729]: nit : I think this is name mangling the functions, right ? We might need a name that indicates it is name mangling the function names. What do others think ? 

cc : @areusch @Mousius

File: src/relay/backend/aot_executor_codegen.cc
Comment (manupa-arm) [inline src/relay/backend/aot_executor_codegen.cc:377]: Why do we need a separate function ? Isnt it possible to improve _CompileEngineLower to accept module name generally ?

File: src/relay/backend/compile_engine.cc
Comment (manupa-arm) [inline src/relay/backend/compile_engine.cc:616]: I think we can have just one Lower function -- is there a reason why mangling the name is bad ?

cc : @areusch

File: src/relay/backend/compile_engine.cc
Comment (manupa-arm) [inline src/relay/backend/compile_engine.cc:621]: Same comment as above

File: src/relay/backend/compile_engine.h
Comment (manupa-arm) [inline src/relay/backend/compile_engine.h:211]: Same as above, Im missing the requirement to have a whole new function.

File: tests/python/relay/aot/test_crt_aot.py
Comment (manupa-arm) [inline tests/python/relay/aot/test_crt_aot.py:353]: Is there a use case, we would not want to name mangle ? Im think name mangled function names are strictly superior. If thats the case, I dont like the idea having to run another pass for all BYOC codegens. 

Seeing the previous conversation with @areusch , I agree that we dont need another PartitionGraphWithModName but also not sure we need to call a seperate pass explicitly either. My suggestion is we should name mangle by default otherwise the compilation pathway feels bit fragile.

At a minimum, we should move this pass as a sub-pass of PartitionGraph and call it by default. 
This comment also has a synergy with why we need LowerWithModName in compile engine. 

I think Im missing whats the value of not performing name mangling when user do provide with a mod_name.

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Hi @giuseros ,

Sorry been busy a bit :) 

I did an initial pass. 

I guess my main concern is why are we duplicating lowering functions with and without name mangling support. I would naturally think user would want to name mangle the function names if a mod_name is provided and mangled names are strictly superior to non-mangled polluted names, irrespective of whether they are being used in AoT or not, unless Im missing something.

I'd like to hear @areusch thoughts as well.

**Comment by giuseros** [comment]
Hi @manupa-arm , 
Thanks for your reply. I pushed a new set of changes trying to address your comments:
* Now I am using name-mangling as default in all TVM
* The `MangleGraph` pass is now the default in the partitioning pipeline
* The mangled mod-name is formed in python and passed down in C++ where we only concatenate to the function names

I guess that, in addition to make the tests pass, we need to agree on naming:
* `MangleGraph`. Is `MangleExternalFunction` better? Because I am only mangling the functions that will be offloaded to the external compiler
* `run_func`: is it ok for everyone if I use only `run`? My problem with only `run` is that it seems a bit too generic. Maybe `run_model`? What do you think?

File: cmake/modules/contrib/VitisAI.cmake
Comment (areusch) [inline cmake/modules/contrib/VitisAI.cmake:24]: why this change?

File: src/relay/transforms/partition_graph.cc
Comment (areusch) [inline src/relay/transforms/partition_graph.cc:483]: this one seems pretty easy to unit test--possible to write one?

File: tests/python/relay/aot/aot_test_utils.py
Comment (areusch) [inline tests/python/relay/aot/aot_test_utils.py:239]: remove

File: tests/python/relay/aot/aot_test_utils.py
Comment (areusch) [inline tests/python/relay/aot/aot_test_utils.py:297]: remove this or the line above

File: python/tvm/relay/backend/utils.py
Comment (areusch) [inline python/tvm/relay/backend/utils.py:21]: can prob not break this line

File: python/tvm/relay/backend/utils.py
Comment (areusch) [inline python/tvm/relay/backend/utils.py:28]: can you catch a more narrow exception?

File: python/tvm/relay/backend/utils.py
Comment (areusch) [inline python/tvm/relay/backend/utils.py:31]: the empty string is valid?

File: python/tvm/relay/backend/utils.py
Comment (areusch) [inline python/tvm/relay/backend/utils.py:34]: can you write unit tests for this function?

File: tests/python/contrib/test_vitis_ai/test_vitis_ai_codegen.py
Comment (areusch) [inline tests/python/contrib/test_vitis_ai/test_vitis_ai_codegen.py:307]: rm

File: python/tvm/relay/transform/transform.py
Comment (areusch) [inline python/tvm/relay/transform/transform.py:732]: per numpydoc style, prefer one-liner for the first line in the docstring

**Review by areusch — CHANGES_REQUESTED** [review]

File: python/tvm/relay/backend/utils.py
Comment (giuseros) [reply python/tvm/relay/backend/utils.py:31]: Yes, if `mod_name` is empty we do `tvmgen_funname`. If it is not empty we do `tvmgen_MODNAME_funname`.

File: src/relay/transforms/partition_graph.cc
Comment (giuseros) [reply src/relay/transforms/partition_graph.cc:483]: The point is that now the pass is integrated inside `the relay._transform.PartitionGraph` pass, so there is now way to test this in isolation. This will be tested by every BYOC test

**Comment by giuseros** [comment]
Hi @manupa-arm , @areusch ,
All the tests have finally passed. Please, let me know if there is anything else I should address.

File: python/tvm/relay/backend/utils.py
Comment (areusch) [reply python/tvm/relay/backend/utils.py:32]: didn't see these, not sure if you had an opinion whether it should be tested?

File: src/relay/transforms/partition_graph.cc
Comment (areusch) [inline src/relay/transforms/partition_graph.cc:504]: nit: s/three/tree/ i think

**Review by areusch — CHANGES_REQUESTED** [review]
just one other nit (honestly mostly care about the unit test, but adding a comment since i saw this)

File: python/tvm/relay/backend/utils.py
Comment (giuseros) [reply python/tvm/relay/backend/utils.py:32]: Yes, sorry, I forgot to push them. I think I added them now.

**Comment by areusch** [comment]
apologies, i think i missed this and it now needs a rebase or merge.

i discussed this with @jroesch a bit as well, who is now working in earnest to merge #7518 in the next two weeks. following that PR, we'll arrive at a state where we can begin to think of the entire compiler as a series of IRModule passes. our discussion was that, at that point, it probably makes sense to move things like name mangling and e.g. the type signature changes proposed by @Mousius to one end of the compiler stack (i.e. at the relay level or at the end of the TIR passes). this way, the middle part of the compiler remains as generic as possible.

we'd like to propose the following:
- let us sync and merge this PR
- we'll land #7518 on top of this in the next two weeks (e.g. aiming for before July 4 US holiday)
- following that PR, we'll likely reorganize this and move it to the end of the TIR passes

to that end, we'd like to briefly prioritize #7518 next week over AOT changes to reduce the churn and let it land, since it's a fairly invasive PR.

if that's ok with you guys, i'll hand this off to @jroesch to merge while i'm ooo next few days.

**Comment by giuseros** [comment]
Hi @areusch , @jroesch , 
I would say, let's get this merged and once #7518 lands, we can sync-up if we want to have the mangling as a pass or as a feature into the compiler. I wouldn't have any strong opinions (and honestly I like the idea of having it as a pass) but maybe @manupa-arm or @Mousius might. 

Thanks,
Giuseppe

**Comment by giuseros** [comment]
Hi @jroesch , @areusch , 
After further thinking we were wondering if we could merge both this PR and #8096 before merging #7518 . This is because #8096 would solve quite a few issues that we have with AoT at the moment. 

What do you think?

**Comment by giuseros** [comment]
A friendly ping @manupa-arm , @jroesch , @areusch !

**Comment by areusch** [comment]
@jroesch can you comment on @giuseros question above?

**Review by areusch — APPROVED** [review]
thanks @giuseros , i think we can merge this now. i'll review #8096 separately, but i think we are okay merging those before #7518 . depending on the complexity of the refactor, we might ask for your help migrating the logic to live in the post-TEcompiler architecture.

**Comment by areusch** [comment]
@manupa-arm @Mousius please take a look and explicitly approve if you're ok w/ this change

**Review by manupa-arm — APPROVED** [review]
LGTM!

**Review by Mousius — APPROVED** [review]
And me :smile_cat:

**Pull request merged**
